rust/lance/src/dataset/scanner.rs

```diff
@@ -2875,6 +2876,60 @@ mod test {
         }
     }
 
+    #[tokio::test]
+    async fn test_column_casting_function() {
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+        let mut data_gen =
+            BatchGenerator::new().col(Box::new(RandomVector::new().named("vec".to_owned())));
+        Dataset::write(data_gen.batch(32), test_uri, None)
+            .await
+            .unwrap();
+
+        let dataset = Dataset::open(test_uri).await.unwrap();
+        assert_eq!(32, dataset.scan().count_rows().await.unwrap());
+
+        let mut scanner = dataset.scan();
+
+        let scan_res = scanner
+            .project_with_transform(&[("f16", "cast_vector_f16(vec)")])
```

Comment:
we can not use the same function but different kernels with different source type?

Reply:
this works for fsl of f32 and f16. Probably good enough for now? It's pretty easy to tweak the UDF to support more

Reply:
ohi i mean can we do `cast_f16` for both f32 or `list[f32]` or `fsl[f32]`? Dont have the "vector" in udf name

Reply:
oh I see. Let's do `cast_list_f16`?

I want to limit how useful this function is as it's opaque to the optimizer. if someone calls `cast_f16(f16)`, it's be a bunch of work to make the optimizer understand that it is no-op

Reply:
if we meant to have this as temporary solution, can we make it clear so? idk whether we can make a `_cast_list_f16` with leading underscore.